app/services/foreman/provision/ssh.rb

```diff
@@ -1,113 +1,176 @@
-      sleep(2)
-      retry
-    rescue Net::SSH::Disconnect
-      logger.debug "Host dropping connections for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::ConnectionTimeout
-      logger.debug "Host timed out for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::AuthenticationFailed
-      logger.debug "Auth failed for #{username} at #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Timeout::Error
-      retry
-    rescue => e
-      Foreman::Logging.exception("SSH error", e)
-    end
+  def upload_template
+    logger.debug "Uploading template to remote system '#{address}'"
+    logger.debug sftp_cmd(mask: true)
+
+    stdout, stderr, status = Open3.capture3(*sftp_cmd)
+
+    logger.debug stdout if stdout.present?
+    log_stderr(stderr)
+
+    return if status.success?
+
+    raise Foreman::Exception.new("Failed to upload '#{@template}' to remote system at #{address}")
+  end
+
+  def passwd_cmd(mask: false)
+    return [] if @options[:password].empty?
+
+    value = mask ? '*****' : @options[:password]
+
+    [{'SSHPASS' => value }, "sshpass", "-e"]
+  end
+
+  def first_cmd(mask: false)
+    ssh_master_opts = ["-o ControlMaster=auto", "-o ControlPath=#{socket_file}", "-o ControlPersist=4m"]
+    [passwd_cmd(mask: mask), "ssh", (ssh_options + ssh_master_opts), address, 'pwd'].flatten
+  end
+
+  def ssh_cmd(host_command, mask: false)
+    [passwd_cmd(mask: mask), "ssh", ssh_options, address, host_command].flatten
+  end
+
+  def sftp_cmd(mask: false)
+    [passwd_cmd(mask: mask), "sftp", ssh_options, "#{username}@#{address}", "<<< $'put #{template}'"].flatten
+  end
+
+  def ssh_options
+    opts = []
+
+    opts << "-o User=#{username}"
+    opts << "-o ControlPath=#{socket_file}"
+    opts << "-o StrictHostKeychecking=no"
+    opts << "-o UserKnownHostsFile=/dev/null"
+    opts << "-o ConnectTimeout=4"
+    opts << "-o Compression=yes"
+    opts << "-o IdentityFile=#{@tmp_key_path}" if @options[:key_data].present?
+
+    methods = []
+    methods << 'publickey' if @options[:key_data].present?
+    methods << 'password' if @options[:password].present?
+
+    opts << "-o PreferredAuthentications=#{methods.join(',')}"
+
+    opts
   end
 
-  def ssh
-    Fog::SSH.new(address, username, options.merge(:timeout => 4))
+  def create_tmp_key
+    return if @options[:key_data].empty?
+
+    file = Tempfile.new("ssh-key-#{uuid}")
+    file.write(@options[:key_data])
+    file.close
+    file.path
+  end
+
+  def check_auth_options
+    return if @options[:key_data].present? || @options[:password].present?
+
+    raise Foreman::Exception.new("Missing :key_data or :password options, no authentication method available")
+  end
+
+  def socket_file
+    @socket_file ||= begin
+      file = Tempfile.new("ssh-socket-#{@uuid}")
+      file.close
+      file.path
+    end
   end
 
-  def scp
-    Fog::SCP.new(address, username, options)
+  def cleanup_files
+    files = [socket_file]
```

Comment:
If I'm reading things right, the connection can still be live at this point (due to controlmaster=4m). Could we issue `ssh $common_option -O exit` to clean up after ourselves rather than leaving it around to exit eventually?

Reply:
Yeah I think we should have cleanup for it as well.

What's the `$common_option` referring to?

Reply:
`-o ControlPath=$path_to_socket $host` should be enough